include/deal.II/lac/solver_idr.h

```diff
@@ -0,0 +1,475 @@
+  VectorType &ghat = *ghat_pointer;
+
+  r.reinit(x, true);
+  v.reinit(x, true);
+  vhat.reinit(x, true);
+  uhat.reinit(x, true);
+  ghat.reinit(x, true);
+
+  // Initial residual
+  A.vmult(r, x);
+  r.sadd(-1.0, 1.0, b);
+
+  // Check for convergent initial guess
+  double res      = r.l2_norm();
+  iteration_state = this->iteration_status(step, res, x);
+  if (iteration_state == SolverControl::success)
+    return;
+
+  // Initialize sets of vectors/matrices whose size dependent on s
+  internal::SolverIDRImplementation::TmpVectors<VectorType> G(s, this->memory);
+  internal::SolverIDRImplementation::TmpVectors<VectorType> U(s, this->memory);
+  internal::SolverIDRImplementation::TmpVectors<VectorType> Q(s, this->memory);
+  FullMatrix<double>                                        M(s, s);
+
+  // Random number generator for vector entries of
+  // Q (normal distribution, mean=0 sigma=1)
+  boost::mt19937               rng;
+  boost::normal_distribution<> nd(0.0, 1.0);
+  boost::variate_generator<boost::mt19937 &, boost::normal_distribution<>>
+    rand_num(rng, nd);
+  for (unsigned int i = 0; i < s; ++i)
+    {
+      VectorType &tmp_g = G(i, x);
+      VectorType &tmp_u = U(i, x);
+      tmp_g             = 0;
+      tmp_u             = 0;
+
+      // Compute random set of s orthonormalized vectors Q
+      // Note: the first vector is chosen to be the initial
+      // residual to match BiCGStab (as is done in comparisons
+      // with BiCGStab in the papers listed in the documentation
+      // of this function)
+      VectorType &tmp_q = Q(i, x);
+      if (i != 0)
+        {
+          for (auto indx : tmp_q.locally_owned_elements())
+            tmp_q(indx) = rand_num();
+          tmp_q.compress(VectorOperation::insert);
+        }
+      else
+        tmp_q = r;
+
+      for (unsigned int j = 0; j < i; ++j)
+        {
+          v = Q[j];
+          v *= (v * tmp_q) / (tmp_q * tmp_q);
+          tmp_q.add(-1.0, v);
+        }
+
+      if (i != 0)
+        tmp_q *= 1.0 / tmp_q.l2_norm();
+
+      M(i, i) = 1.;
+    }
+
+  double omega = 1.;
+
+  bool early_exit = false;
+
+  // Outer iteration
+  while (iteration_state == SolverControl::iterate)
+    {
+      ++step;
+
+      // Compute phi
+      Vector<double> phi(s);
+      for (unsigned int i = 0; i < s; ++i)
+        phi(i) = Q[i] * r;
+
+      // Inner iteration over s
+      for (unsigned int k = 0; k < s; ++k)
+        {
+          // Solve M(k:s)*gamma = phi(k:s)
+          Vector<double> gamma(s - k);
+          {
+            Vector<double>            phik(s - k);
+            FullMatrix<double>        Mk(s - k, s - k);
+            std::vector<unsigned int> indices;
+            unsigned int              j = 0;
+            for (unsigned int i = k; i < s; ++i, ++j)
+              {
+                indices.push_back(i);
+                phik(j) = phi(i);
+              }
+            Mk.extract_submatrix_from(M, indices, indices);
+
+            Mk.gauss_jordan();
```

Comment:
... and this more expensive anyway.

Reply:
I will change to use `FullMatrix::invert` which hard codes up to 4x4 inverse, then uses `gauss_jordan()` if it is larger.